### PR TITLE
GEM vfat dataformat update v301

### PR DESCRIPTION
**Pull request by jshlee**
#### PR description:

* Updating gem dataformat to v301 as documented in https://docs.google.com/spreadsheets/d/1iKMl68vMsSWgr8ekVdsJYTk7-Pgy8paxq98xj6GtoVo/edit#gid=929454128
* more HW status bits added, mainly to unused bits
* added initialisers for constructors 

#### PR validation:

* runTheMatrix ran ok
* no changes for gem

### CODE REVIEW

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
+1

File: EventFilter/GEMRawToDigi/interface/AMCdata.h [hunk 1]
```diff
@@ -8,54 +8,63 @@ namespace gem {
...
   union EventHeader {
     uint64_t word;
     struct {
-      uint64_t ttsState : 11;   // GLIB TTS state at the moment when this event was built.
-      uint64_t davCnt : 5;      // Number of chamber blocks
-      uint64_t buffState : 24;  // buffer error
-      uint64_t davList : 24;    // Bitmask indicating which inputs/chambers have data
+      uint64_t ttsState : 4;  // GLIB TTS state at the moment when this event was built.
+      uint64_t
+          pType : 4;  // Payload type: can be one refering to different zero suppression schemes in normal data taking or calibration type. Note in calibration type the entire GCT (GEM Chamber Trailer) is skipped
```
Comment (perrotta) [inline EventFilter/GEMRawToDigi/interface/AMCdata.h]: If you split the comment into different lines you can avoid clang format splitting the declararation part

File: EventFilter/GEMRawToDigi/interface/AMCdata.h [hunk 2]
```diff
@@ -8,54 +8,63 @@ namespace gem {
...
+      uint64_t davList : 24;    // Data Available list: a bitmask indicating which chambers have data in this event
     };
   };
   union EventTrailer {
     uint64_t word;
     struct {
-      // GLIB is out‐of‐sync (critical): L1A ID is different for different chambers in this event (1 bit)
-      uint64_t oosGlib : 40;
-      uint64_t chTimeOut : 24;  // GLIB did not receive data from a particular input for this L1A
+      uint64_t
```
Comment (perrotta) [inline EventFilter/GEMRawToDigi/interface/AMCdata.h]: If you split the comment into different lines you can avoid clang format splitting the declararation part

File: EventFilter/GEMRawToDigi/interface/VFATdata.h [hunk 3]
```diff
@@ -12,10 +12,11 @@ namespace gem {
       uint64_t msData1 : 16;  ///<channels from 65to128
       uint64_t bc : 16;       ///<Bunch Crossing number, 16 bits
       uint64_t ec : 8;        ///<Event Counter, 8 bits
-      ///<normally 0x1E. 0x5E indicates that the VFAT3 internal buffer is half-full, so it's like a warning
-      uint64_t header : 8;
-      uint64_t crcCheck : 8;  ///<bits 183:177 are not used, should be 0, bit 176 is 1 if CTP7 detected a CRC mismatch
-      uint64_t pos : 8;       ///<an 8bit value indicating the VFAT position on this GEB (it can be 0 to 23)
+      uint64_t
+          header : 8;  ///<normally 0x1E. 0x5E indicates that the VFAT3 internal buffer is half-full, so it's like a warning
```
Comment (perrotta) [inline EventFilter/GEMRawToDigi/interface/VFATdata.h]: If you split the comment into different lines you can avoid clang format splitting the declararation part

File: EventFilter/GEMRawToDigi/interface/AMCdata.h [hunk 4]
```diff
@@ -8,54 +8,63 @@ namespace gem {
...
+          BCL : 4;  // 1st bit, BC0 locked - If 0, this is a bad condition indicating a problem in the clock or TTC command stream (critical condition)
+      uint64_t DR : 1;  // DAQ Ready - If 0, this means that AMC13 is not ready to take data (critical condition)
+      uint64_t CL : 1;  // DAQ clock locked- If 0, this indicates a problem in the DAQ clock (critical condition)
+      uint64_t ML : 1;  // MMCM locked - Should always be 1
+      uint64_t BP : 1;  // Backpressure - If this is 1, it means that we are receiving backpressure from AMC13
+      uint64_t
+          oosGlib : 40;  // GLIB is out‐of‐sync (critical): L1A ID is different for different chambers in this event (1 bit)
```
Comment (perrotta) [inline EventFilter/GEMRawToDigi/interface/AMCdata.h]: The struct doesn't add up to 64 bits any more
If I read correctly the documentation, ```oosGlib``` should be 32 bits only

File: EventFilter/GEMRawToDigi/interface/GEBdata.h [hunk 5]
```diff
@@ -11,34 +11,47 @@ namespace gem {
   union GEBchamberHeader {
     uint64_t word;
     struct {
-      uint64_t inputStatus : 23;      // Input status (critical) only first 13 bits used
-      uint64_t vfatWordCnt : 12;      // Size of VFAT payload in 64bit words expected to send to AMC13
-      uint64_t inputID : 5;           // GLIB input ID (starting at 0)
-      uint64_t zeroSupWordsCnt : 24;  // Bitmask indicating if certain VFAT blocks have been zero suppressed
+      uint64_t BxmVvV : 11;           // 1st bit BX mismatch VFAT vs VFAT
+      uint64_t BxmAvV : 1;            // BX mismatch AMC vs VFAT
```
Comment (perrotta) [inline EventFilter/GEMRawToDigi/interface/GEBdata.h]: In the comment of the previous version of the dataformat it was written that "only first 13 bits used". But you are here rewriting the 12th and 13th bit: is the new dataformat still backward compatible as such?

**Comment by perrotta** [comment]
Is the new version of the dataformat still backward compatible with the old one? I spotted at least one case in the ```GEBchamberHeader``` where the new field could overwrite meaningful bits in the previous dataformat (see https://github.com/cms-sw/cmssw/pull/27419#discussion_r300943104), but I did not check explicitely all other modified fields.
Did you check that you are still able to correctly unpack old data (e.g slice test)? Could you please suggest a test to verify it?

File: EventFilter/GEMRawToDigi/interface/GEBdata.h (hunk 5 above)
Comment (jshlee) [reply EventFilter/GEMRawToDigi/interface/GEBdata.h]: The data format is compatible. Before the inputStatus was just lumped together.

File: EventFilter/GEMRawToDigi/interface/AMCdata.h (hunk 4 above)
Comment (jshlee) [reply EventFilter/GEMRawToDigi/interface/AMCdata.h]: fixed

**Comment by jshlee** [comment]
I will check with slices test data. However, the inputStatus var is not used for unpacking, only for checking geb status.

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by jfernan2** [comment]
hold

It needs to fix dot in Histo file name on:
https://github.com/cms-sw/cmssw/blob/a2816f87856da3054918a05646e4bb97067b5520/DQM/GEM/plugins/GEMDQMStatusDigi.cc#L87

**Comment by jshlee** [comment]
@jfernan2 - fixed

**Comment by jfernan2** [comment]
unhold

**Comment by jfernan2** [comment]
please test

**Comment by perrotta** [comment]
@jshlee please let us know when you have completed the check anticipated in https://github.com/cms-sw/cmssw/pull/27419#issuecomment-509270959 , and you can then post here evidence that we will still be able to correctly read the geb status also with data that were recorded with the previous dataformat

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1
- Code changes in code reflect the new v301 GEM dataformat
- Mostly previously unused bits are now overwritten in the new format.
- There are a few bits which change meaning in v301 wrt v300: according to @jshlee (see [presentation](https://indico.cern.ch/event/834029/contributions/3495120/attachments/1879931/3097059/gemreco.pdf) at the reco meeting on Jul 12), they are only meant to monitoring and DQM.
- As such, GEM DQM code will have to be updated to cope with the new and old data formats; as of now, possibly offending code has been commented out [here](https://github.com/cms-sw/cmssw/pull/27419/files#diff-092b9980cd4579fea8c6ade32e93eda8R122).
- Jenkins tests pass and show no difference, as they should

**Comment by perrotta** [comment]
For the record, there are overwritten fields in the following ```struct```'s:
  - CDFTrailer
  - EventHeader
  - EventTrailer
  - GEBchamberHeader
  - GEBchamberTrailer

**Comment by fabiocos** [comment]
@jshlee I would like a clarification about the implication of these data format changes: from the comment https://github.com/cms-sw/cmssw/pull/27419#issuecomment-510928150 I am not sure the code is at present compatible with existing data, and how this should be managed in case. Could you please comment about this point?

**Comment by fabiocos** [comment]
@jshlee from the discussion I understand that there is no implication but for part of the DQM. So I will move forward with the integration

**Comment by jshlee** [comment]
@fabiocos - thanks. I was at a conference and couldn't follow this thread.

**Comment by fabiocos** [comment]
+1

**Pull request merged**